### ISSUE TITLE
HDMI screen "pickiness"

### ISSUE THREAD

**Issue by Sch-LikA** [hoglet67/RGBtoHDMI#170]
I just built a small batch of RGBtoHDMI, that is a really amazing project ! First a huge thanks.

At first I thought I screwed up my build since nothing showed on screen. After lots of testing on 7 different screens :

- Dell 2407 over hdmi-to-dvi
- Dell 2410 over hdmi and "hdmi-to-dvi"
- Samsung 23h9x over hdmi-to-dvi
- Nexdock (would be so cool) hdmi
- standalone RTD2660H controller + 11' screen hdmi
- HP LP1965 hdmi-to-dvi
- Dell U2413 hdmi

Only the last two work. At least they work fine :-)

Anything I can do to help support the other ones ?

**Comment by hoglet67** [comment]
I can't find a reference on the internet to a samsung 23h9x. Can you provide links to the specs for this?

There seems to be compatibility issues between the NexDock and the Pi:
https://www.raspberrypi.org/forums/viewtopic.php?t=245153

Have you tried booting with SW1 pressed? This should force 60Hz output.

What source are you using?

**Comment by Sch-LikA** [comment]
Sorry wrong Samsung model, it is a 2343BW.

I have a "2nd release" Nexdock which works fine with about anything I ever threw at it (Raspberry 0,1,2,3 & 4, MiSTerFPGA, Smartphones). I just tried with SW1 pressed, not better, and tried connecting another laptop on the USB in case the Nexdock used USB to detect activity, no better result.

Just tried with SW1 pressed and got my 2410 working, great. I am now using Auto@50-60Hz (it passes the 50Hz test btw ...) I will try the other screens later this week-end. Would Auto@50-60Hz be a better default choice ?

My source is a C128 (DCR) 80 columns, no TTL board. Works fine, I tried the "VDC Mode Mania" "demo", which does a lot of tricks to get higher resolutions & colors on screen, I had to set "Auto Switch" to off since the interlaced modes where troubling rgb-to-hdmi sync. It works then, but it is quite strange to see a TFT flicker :-)

Thanks for the support !

**Comment by IanSB** [comment]
>Just tried with SW1 pressed and got my 2410 working, great. I am now using Auto@50-60Hz (it passes the 50Hz test btw ...) I will try the other screens later this week-end. Would Auto@50-60Hz be a better default choice ?

I'm not sure what's going on there. The default resolution of "Default@EDID" is the same as "Auto@50-60Hz" except that it tries to be safer and always sends 60Hz unless the EDID indicates that 50Hz will work whereas Auto@50-60Hz always sends 50Hz for 50Hz sources which means you will get blank screens on monitors that don't support 50Hz when using that.

Did you change the resolution setting to something else before running your monitor tests? (holding SW1 sets it back to "Default@EDID" as well as temporarily forcing 60hz)

Also be aware that a lot of old Dell monitors will work with 50Hz input and thus pass the test but they do an internal 50Hz to 60Hz conversion and send video to the LCD at 60Hz which makes their 50Hz support useless as you end up with the same judder on smooth motion as you would get when sending 60Hz out of the Pi instead. 

My Dell 2707 and 2007 both do this but they do actually produce a display so I'm surprised you are getting no output with your 2407 which is from the same year.

HP monitors of the same vintage like your LP1965 and the LP2065 are much better as they actually work at 50Hz with no internal conversion.

> I tried the "VDC Mode Mania" "demo",

Non standard modes are not yet fully supported on the C128 but at least some of them now work.
There is a beta test version on my github fork which you can try with "VDC Mode Mania"
https://github.com/IanSB/RGBtoHDMI/releases/tag/7eef881

This build also has a "HDMI mode" setting that allows the HDMI output mode to be changed which might also have an effect on some of the non-working monitors although you will have to change the setting on a working monitor to start with.
Note changing from DVI compatible will stop it working with the HP LP1965 and probably some of your other DVI monitors

**Comment by Sch-LikA** [comment]
Hi again,

I just retried my monitors, with other cables this time, and yes, they all work. Sorry for the wrong alarm. The only problem remaining is the Nexdock, it still refuses to show any image.

I will try your new build next and run "VDC Mode Mania" !

**Comment by Sch-LikA** [comment]
OK amazing stuff. Your beta firmware works fine on my Dell 2410.

- Interlaced modes are still flickering but sync seems stable by default, much better already
- vdc-fli renders amazingly well
- vdc-hfli still has issues on my setup, sync is not stable and image cannot be seen correctly
- vdc-himono gives "auto switch : no profile matched" and does not work well
- Big-text-pal & Big-text-ntsc gets recognized ok but shows a kind of wave effect on the right side of the screen, still readable 

I will get my CM8833 (mkI) I just repaired out to compare a few colors, but it is much better now. To avoid flicker, would some kind of Amiga 2024 "high persistence phosphor" emulation help (no idea how to do that maybe mix the two frames in a buffer and render around 50hz ?) Thanks again !!

**Comment by IanSB** [comment]
>"VDC Mode Mania"

Unfortunately I don't have a Commodore 128 (yet) so I'm relying on feedback from third party testers

>  Interlaced modes are still flickering but sync seems stable by default, much better already

The problem with those two interlaced modes (VDC-IHFLI 266 & VDC-ITFLI 315) is that they output progressive syncs with interlaced video so effectively they are alternating progressive frames for a dithering effect and they won't be weave deinterlaced because there is no way to detect that is what they are doing. If they had interlaced syncs that would be detected and they would be weave deinterlaced. It might be possible to have a forced deinterlace mode but then the pair order of the woven lines would be random each time the mode was selected.

> vdc-himono gives "auto switch : no profile matched" and does not work well

Not sure what mode you mean, it supports VDC-IMONO 749i & VDC-IM800 651i which I've been told work OK and are deinterlaced correctly as they output interlaced syncs. It looks like you are using another mode so what version of the demo are you running? 

> Big-text-pal & Big-text-ntsc gets recognized ok but shows a kind of wave effect on the right side of the screen, still readable

This probably means the line length is set incorrectly in the geometry menu as that will cause vertical noise bars so try adjusting that setting. When adjusting, you are looking for a result that is either all clean or all noise, not alternating noisy and clean bars and after you get a result, tweak the sampling phase in the sampling menu to get rid of any remaining noise and save the profile.
If you get a good result, please let me know the values you changed.

**Comment by Sch-LikA** [comment]
I am using V1.01 of Vdc Mode Mania. VDC-IMONO is 720x700 interlace monochrome, refresh at ~21Hz according to the demo. Interestingly when I press button 1 the image looks half ok behind, but when out of the menu I see the led blinking and the image appears interlaced flickering with the led.

I just set "line length" to 1017 and screen is perfect in both Big-text-* modes. I played with sampling phase, all but 3 and 4 work fine. So 1017 as line length and anything but 3 or 4 as sampling phase works fine on my setup !

That stuff is amazing. I just built an analog module for my CBM-II, I'll try next. Thanks for your time !

**Comment by Sch-LikA** [comment]
oh, btw this is on a PAL C128DCR.

**Comment by IanSB** [comment]
>I just set "line length" to 1017

To confirm that is the correct setting: the original noisy value was 1016, does setting it to 1018 produce noise as well?

>VDC-IMONO is 720x700 interlace monochrome, refresh at ~21Hz according to the demo

Can you post a screencap of the Source Summary menu in the info menu when running that mode.
(Go into that menu then press the up and down buttons and that will screencap the menu image to the SD card as a PNG)

**Comment by IanSB** [comment]
@Sch-LikA 

> I am using V1.01 of Vdc Mode Mania.

My other tester has been using V1.11 of VDC Mode Mania and it works with that (apart from VDC-HFLI_403 which has very strange results in the logs and may have out of spec H or V sync pulses)

That appears to be downloadable here:
https://csdb.dk/release/?id=161195
Can you try that version?

**Comment by Sch-LikA** [comment]
Hi ! I just tried Vdc Mode Mania 1.11 

Mode 1 & 2 (VDC-IHFLI & VDC-ITFLI)  : All ok
Mode 3 : Pictures have a bad vertical alignment, picture starts at ~1/3 top and there is a black line in the middle, like in this capture : 
![capture2](https://user-images.githubusercontent.com/44978559/105361386-edd3d800-5bf9-11eb-99d1-b615880c23b4.png)

Mode 4 - VDC-HFLI 403 : Gets detected but does not sync
Mode 5 : Works fine !

Here is the capture of VDC 1.01 IMONO mode :
![capture0](https://user-images.githubusercontent.com/44978559/105361244-c8df6500-5bf9-11eb-9a40-107a0a46079c.png)


Text modes : I confirm 1017 is the correct line length, 1016 and 1018 produce noise.

**Comment by IanSB** [comment]
@Sch-LikA 
>Here is the capture of VDC 1.01 IMONO mode

The current version of IMONO mode has a PAL line length of 1024 but your version has an NTSC line length of 1016 instead.

I think I've figured out what the problem is with the other modes so there should be a fix soon

**Comment by IanSB** [comment]
@Sch-LikA 

Here is an updated beta 14 which might work properly in all VCD modes:

https://github.com/IanSB/RGBtoHDMI/releases/download/98d62b35/beta14.zip

**Comment by IanSB** [comment]
@Sch-LikA 

Now up to beta 16 which according to my other tester works with all VCD Mode mania:

https://github.com/IanSB/RGBtoHDMI/releases

One option that requires manual override is NTSC Big text because it shares the same timing as regular NTSC.
To select that you have to change Auto Switch to off and then select the NTSC text sub-profile (not profile).
I may implement some sort of popup menu for circumstances like this where two or more sub-profiles share the same timing.

**Comment by Sch-LikA** [comment]
Just tried b16, Vdc Mode Mania 1.01 works fine in all modes.

Vdc Mode Mania 1.11 still has the same issue in mode 3 (the astronaut), and mode 6 (VDC-IM800) flickers and does not sync. All the rest works fine !

I even had another issue resolved after forcing a resampling, I built a 4 option roms switcher, and Basic8 was having issue in text mode. It does now work fine. I'll try some Basic8 graphic modes later.

Thanks a lot, huge progress.

**Comment by IanSB** [comment]
@Sch-LikA 

Please post source summary screencaps for those that don't work.

**Comment by Sch-LikA** [comment]
Hi ... sorry for delay, here are the screenshots :

![capture9](https://user-images.githubusercontent.com/44978559/108120206-379cca80-70a1-11eb-8846-d08b01f75b80.png)
![capture10](https://user-images.githubusercontent.com/44978559/108120210-38356100-70a1-11eb-98d0-767a8d119645.png)
![capture11](https://user-images.githubusercontent.com/44978559/108120211-38cdf780-70a1-11eb-83cf-1d58c92f47f3.png)
![capture12](https://user-images.githubusercontent.com/44978559/108120213-38cdf780-70a1-11eb-83e8-09e14b13cf7b.png)
![capture13](https://user-images.githubusercontent.com/44978559/108120214-38cdf780-70a1-11eb-9bc6-8941a2e953b1.png)
![capture14](https://user-images.githubusercontent.com/44978559/108120215-39668e00-70a1-11eb-96d1-7cc1f16ea861.png)
![capture15](https://user-images.githubusercontent.com/44978559/108120218-39ff2480-70a1-11eb-9c29-3f141ddaafda.png)
![capture16](https://user-images.githubusercontent.com/44978559/108120219-39ff2480-70a1-11eb-9995-33691d38e58c.png)

And finally text mode of Basic8 does blink too :

![capture17](https://user-images.githubusercontent.com/44978559/108120264-4a170400-70a1-11eb-8768-74757ac8e37e.png)

Btw I was working on Atari ST's lately, and was wondering if the monochrome hires "kindof" vga (640x400, 72hz vert /31.5khz hor) could be supported by RGBtoHDMI ?

**Comment by IanSB** [comment]
@Sch-LikA 

>Btw I was working on Atari ST's lately, and was wondering if the monochrome hires "kindof" vga (640x400, 72hz vert /31.5khz hor) could be supported by RGBtoHDMI ?

Already works, see:
https://github.com/c0pperdragon/Amiga-Digital-Video/issues/6

**Comment by Sch-LikA** [comment]
Amazing work guys, thanks !

**Comment by victokra** [comment]
Hi, tokra here the author of "VDC Mode Mania". I just got an RGB2HDMI and am trying my modes with it. I had success with all of them but ITFLI, which is kind off picky and results in a flashing screen. Still investigating this.

> The problem with those two interlaced modes (VDC-IHFLI 266 & VDC-ITFLI 315) is that they output progressive syncs with interlaced video so effectively they are alternating progressive frames for a dithering effect

This is news to me :) I did not consciously do anything else with these two modes than with the monochrome-modes. They both set the VDC-register 8 to "interlaces video & sync". Maybe adding colour to the mix makes the VDC do strange things? Would be interested to go deeper into this if someone can explain what exactly is and should be happening.

Anyway, the wrongly placed picture of VDC-FLI in V1.11 is due to a small screw-up of mine I realized only years later. Other than earlier versions, I must have copied a line wrong and now it load the settings for the wrong mode and as such sets reg 7 (VSYNC-position) wrong. It is probably the easiest to just use V1.13 from my homepage which fixes this issue and speeds up the screen-copy and screen-clean a bit since it now uses assembler-routines. Get it here: https://www.tokra.de/c128/vdcmodemania-v113.zip

**Comment by IanSB** [comment]
@victokra 
>I had success with all of them but ITFLI, which is kind off picky and results in a flashing screen. Still investigating this.

Flashing screen either means there isn't a profile that matches the timing or the selected profile has a capture size that exceeds the size of the video frame. (See the help menus in the latest release)

>>The problem with those two interlaced modes (VDC-IHFLI 266 & VDC-ITFLI 315) is that they output progressive syncs with interlaced video so effectively they are alternating progressive frames for a dithering effect

>This is news to me :) I did not consciously do anything else with these two modes than with the monochrome-modes. They both set the VDC-register 8 to "interlaces video & sync". Maybe adding colour to the mix makes the VDC do strange things? Would be interested to go deeper into this if someone can explain what exactly is and should be happening.

The main problem is that I don't have any Commodore 128 hardware so debugging issues like this is not easy.

What does the source summary show for those modes (i.e are they being detected as progressive or interlaced?)
Go into the info menu and select source summary while those modes are being displayed. 
You can screencap the menu by pressing SW2 & SW3 together.

>I have attached my profiles for Mode Mania and the newer modes.

Can all these mode files be included in the single Commodore 128 profile folder?
(i.e. do they all have different timings so they can be auto selected?)

**Comment by IanSB** [comment]
@victokra 
>These are the screenshots of the menus. 

Progressive / interlace detection is done by first measuring the average line duration over 100 lines of video then measuring the time for two frames / fields. This value is then divided by the line duration (with suitable rounding) to find the number of lines for two frames /  fields. 
If this number is odd then the video is interlaced else it is progressive.
Could you use the info menu and select the save log and EDID option while displaying a problem mode then post log.txt from the SD card as this should show more details about those measurements.

**Comment by victokra** [comment]
Here you go with the log. Hope you can make something out of it. Searching for IHFLI and ITFLI should bring you to the affected sections.

[Log.txt](https://github.com/hoglet67/RGBtoHDMI/files/11567593/Log.txt)

**Comment by IanSB** [comment]
@victokra 

```
Autoswitch match: VDC-IHFLI_266 (8) = 63182, 63817, 266, 3
Lines per frame = 266, (266)
Actual frame time = 16891206 ns (non-interlaced), line time = 63500 ns

Autoswitch match: VDC-ITFLI_315 (12) = 63680, 64320, 315, 3
Lines per frame = 315, (315)
Actual frame time = 20160203 ns (non-interlaced), line time = 64000 ns
```
The time for two frames/fields is only logged when interlace is detected and in the above cases the value for only one frame is logged instead when progressive is detected but this value is merely the two frame value divided by 2 so it is easy to reproduce the original value by multiplying by two
```
16891206 * 2 = 33782412 / 63500 = 532.0064 so 532 lines
20160203 * 2 = 40320406 / 64000 = 630.0063 so 630 lines
```

As these are both even numbers I don't think they can be interlaced signals as there has to be half a line extra per field for interlace. Maybe some combination of VDC register settings is stopping interlaced sync from being generated?

Do the other interlaced modes work OK?

**Comment by victokra** [comment]
Yes, both VDC-IMONO (720x700) and VDC-IMONO800 are recognized as interlaced and displayed correctly. I assume it is the color-activation that messes up the interlace. The VDC was known to be a buggy chip. I found another interlace color-program (IPAINT) that shows exactly the same problem. Anyways, thanks for the explanation. That gives me a good starting point to investigate and experiment.

Maybe it could be made an option to force interlace and pick top / bottom field order?

**Comment by IanSB** [comment]
@victokra 

Here is a proposed Commodore 128 profile set with your changes plus I also set the frame buffer to double height in all profiles as that is usually required for variable scanlines and may also help with interlaced display

Delete the existing Commodore_128-80 folder in /profiles and replace with the attached folder.

Also delete the Commodore_128-80 folder in /saved_profiles as that will override the new values in /profiles

I have set the video to progressive for VDC-IHFLI_266 & VDC-ITFLI_315 but you might want to change back to interlaced to see if the double height framebuffer setting makes any difference.

If the above looks OK I'll add it to the next release, if not please post your corrections.
[Commodore_128-80.zip](https://github.com/hoglet67/RGBtoHDMI/files/11570283/Commodore_128-80.zip)

**Comment by victokra** [comment]
Ok, these profiles work just as well, but did not improve the interlace-recognition. I investigated a little more why the monochrome modes are recognized correctly and the color ones are not. Basically the difference is in the character-height register (reg 9) which has to be set for the character-height - 1. While for monochrome-interlaced-modes these are set to a height of 7 (odd) for the color modes they are set to 4 or 6 (even). I've checked that odd values always lead to interlace being detected correctly and even values always don't. I can see no really good reason for this though. As said the VDC was known to be buggy.

The number of lines is calculated by (reg 4 +1) * (reg 9 + 1).
For IHFLI this is: (132 +1) * (3 +1) = 532 interlaced
ITFLI: (104 +1) * (5 +1) = 630 interlaced
IMONO: (106 +1) * (6 + 1) = 749 interlaced
IM800: (92 + 1) * (6 +1) = 651 interlaced

However for the color-modes I **have to** set to the register to even values as each field has to have an even number of lines for color. E.g. with 4 I get the following colors:

AAAAAAAA
BBBBBBBB
AAAAAAAA
BBBBBBBB

CCCCCCCC
DDDDDDDD
CCCCCCCC
DDDDDDDD

A nice setup. But with 5 the setup would look like this:

AAAAAAAA
BBBBBBBB
AAAAAAAA
BBBBBBBB
AAAAAAAA

CCCCCCCC
DDDDDDDD
CCCCCCCC
DDDDDDDD
CCCCCCCC

So, less color-granularity (2 colors per 5 lines instead of 2 per 4 lines). I've actually tried such a setting and it was correctly recognized, but it just does not make much sense :)

Only option seems to force interlace for even numbered line count and to choose field order.

**Comment by IanSB** [comment]
@victokra 

Have you tried adjusting register 5 (Vertical Total Fine Adjust) ?
http://www.ffd2.com/fridge/chacking/c=hacking2.txt

If you have one of those cheap USB logic analysers it might be interesting to look at the relationship between the H and V sync pulses on two adjacent frames in progressive, normal interlaced and bad interlaced modes.

**Comment by victokra** [comment]
Yes, but sadly contrary to original documentation register 5 only works in NON-interlaced mode. Since I saw the first examples of interlace on the VDC back in 1992 I've always wondered why they always set register 5 to the same value as register 9 (character height). It really makes no sense from a technical point of view. But only if both registers have the same value interlaced display is working at all. You would think that lower values in register 5 would be the fine adjust of the vertical total number of lines, but then interlace is just lost completely and only one field is displayed (also on CRTs). The fine adjust in reg 5 only works as documented in NON-interlace mode and in interlace-mode does not add any lines to VerticalTotal at all. I've also checked the few "professional" interlace software from the past (like IPaint and Graphic Booster) and they all set reg 5 to the same value as reg 9 - with the above mentioned side-effect that you get an even total number of lines with the wrong combination which the RGB2HDMI does not recognize as Interlace. Which worked fine on the Commodore-monitors I use (1901 and 1084), they don't care about the missing half extra line in each interlaced-field.

To get "correct" interlace the value in register 9 multiplied by the total number of lines in register 4 must be odd. This is the case for my VDC-IMONO (107*7=749 total vertical lines) and VDC-IM800 (93*7=651 total vertical) and that's why auto-deinterlace with the RGB2HDMI. Xander Mol did a video of VDC Mode Mania with his HDMI-capture devices which you can check here:

https://www.youtube.com/watch?v=toeOYcWU-6o

The color-modes flicker similar to as they would on a CRT because the interlace is not recognized. The monochrome-interlace modes starting a 3:54 are rock sold in comparison.

I still chalk this up to the VDC being a buggy chip. Reading about its history its a wonder it works as well as it does.

I do not have a logic analyzer to do further investigations, but will ask around :) Thanks for your suggestions.

**Comment by IanSB** [comment]
@victokra 
>Those worked fine on the Commodore-monitors I use (1901 and 1084), they don't care about the missing half extra line in each interlaced-field.

If they are sending progressive syncs then they shouldn't really work fine on an analog monitor as it would draw both sets of images in the same place rather than interleaving them but on such a small monitor with the flicker it can be difficult to see that.

The only other possibility is that these non standard modes are doing something else with the syncs which is not detected by current RGBtoHDMI code but as mentioned above the only way to determine that is to use a logic analyser or digital storage oscilloscope to look at H & V syncs on two adjacent fields
i.e. something like this:
https://www.ebay.co.uk/itm/314272582941
(There are many sellers)

**Comment by victokra** [comment]
I only can say for sure that I visually ascertained on my Commodore 1901 which field came first back when I programmed the demo back in 2011. If I flipped the field order around in the display routine the picture looked wrong.

Sorry, I have no idea about logic analyzers and would not even know how and where to connect them and what to look for when running the software or how to set it up.

I agree that getting to the bottom of this quirk is the final goal, but would a manual override be possible in the meantime?

**Comment by IanSB** [comment]
@victokra 
>If I flipped the field order around in the display routine the picture looked wrong

Then there must be something different between the syncs in each field. The only thing I can think of is that the vsync is shifted by half a line in one field which would make the first field one line longer than the second field.
e.g. IHFLI would be 266.5 & 265.5
It would still be possible to discriminate between fields if that was the case (which is what your analog monitor appeared to be doing).
Here is a version of kernelrpi.img that forces interlace to true for all inputs. (This may mess up other modes)

[kernelrpi-force-interlace.zip](https://github.com/hoglet67/RGBtoHDMI/files/11586956/kernelrpi-force-interlace.zip)


Overwrite your existing copy on the SD card and reboot
Note you will have to change the video type from progressive back to interlaced in the two problem profiles for this to have any effect.

There are several possible outcomes from this change:
1. no effect
2. it works!
3. It works but the fields are reversed
4. It works but the fields are offset by a line or two
Post a screencap if it makes a difference
If this does make a difference then I could add a force interlace setting to the Vsync Type to always assume interlace just for those profiles. (The current interlace setting just allows interlace or progressive as detected)

**Comment by victokra** [comment]
2. It works!

![capture10](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/20f96826-62a9-4fa2-8361-924b58e98ea8)
![capture14](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/0e73e5e3-50ea-4251-849d-c380c721ebbc)

Not sure why the capture of ITFLI is that stretched. It looks fine on the display. Original resolutions are 640x480 for the Terminator-pic and 640x576 for the train.

So, this indeed seems to be the solution for now. Thanks for your efforts! Hope to see this implemented in the next release/beta.

**Comment by IanSB** [comment]
@victokra 
>2. It works!

That's great although it confirms the output is very non-standard as no video signal like that should have different line numbers in each field.

Anyway here is an updated set of files:
[C128_even_interlace.zip](https://github.com/hoglet67/RGBtoHDMI/files/11593931/C128_even_interlace.zip)

Replace kernelrpi.img and the Commodore_128-80 profile set on your SD card and reboot
Make sure to delete any saved profiles as they will override the new ones.

This version has a new setting in Vsync Type of "Force Interlaced" which forces interlaced detection always for that profile.
The new Commodore_128-80 profile has the VDC-IHFLI & VDC-ITFLI sub-profiles updated to use that setting. (I also renamed them as VDC-IHFLI_532i & VDC-ITFLI_630i)

If you can test that and confirm everything is working as expected then I'll commit that to the repository for the next beta.

BTW I've been working with c0pperdragon to do an RGBtoHDMI adapter for the Commodore 64 video. It uses a small plug in board which sits under the VIC II to pick up the digital signals going to the VICII chip and it then sends them out of the C64 on the UHF modulator's phono in an encoded format so that an RGBtoHDMI can plug into that and decode the data to make a completely noise free bit perfect image. 
Do you think there would be interest in a C128 variant of that?

**Comment by victokra** [comment]
Thanks. Here's my version

[Commodore_128-80.zip](https://github.com/hoglet67/RGBtoHDMI/files/11594403/Commodore_128-80.zip)

I made the following changes:

- adjusted the vertical offset for VDC-IHFLI_532i.txt
- deleted Cmd_128_60Hz_262.txt - the ROM-sets by Commodore only produces 264 lines in NTSC. Not sure where the 262 lines alternative came from. Feel free to keep it if there's another good reason.
- added Cmd_128_50Hz_320.txt - there was a ROM-revision by Commodore in 1986 which adjusted PAL-settings for VDC to be more correct (15625 Hz and 312 lines). However all machines before that had slightly off PAL-timing (1016 line-length = 15748 Hz, 320 lines). I think about 50% off all PAL-C128 out there will be the older revision.
- renamed Cmd_128_50Hz.txt  to Cmd_128_50Hz_312.txt to reflect above change
- picked palette RGBI for VDC-Spectrumania-4in1_410.txt - this one requires black yellow, was set to RGBI (CGA) before

Regarding c0pperdragon: I understand this basically reproduces the VIC-chip's output on its own? Does the mod require the HF-modulator? A lot of people have removed theirs from their machine for other picture-improving hardware (on C64 and C128) to remove "jailbars". If just the jack is used that should be fine. C128 folks would be just as interested I think. Not sure if the extra registers of the VIC-IIe in the C128 matter (test-bit and 2 Mhz-mode).

**Comment by IanSB** [comment]
@victokra 
>I made the following changes:

Thanks for your updates.

>deleted Cmd_128_60Hz_262.txt - the ROM-sets by Commodore only produces 264 lines in NTSC. Not sure where the 262 lines alternative came from. Feel free to keep it if there's another good reason.

The NTSC version of PETSCII robots produces 262 line video and it was added to support that, see: https://github.com/hoglet67/RGBtoHDMI/issues/266

The old c0pperdragon board was a multi board YUV adapter, this new board only works with RGBtoHDMI and just fits under the VICII, monitors the data bus and generates a signal called "lumacode" (a four level analog mono video signal at double the normal clock rate) which can be output on a phono connector such as the RF one (RF output disconnected). An analog RGBtoHDMI can decode that back into a C64 image. (It's a way of getting bit perfect video out of the computer on a single wire).

![vicIIboard1](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/3de070b3-3afa-40b7-bdf3-8870b150780b)
![vicIIboard2](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/7523f919-cd38-454d-ace8-33aa52e74993)
![capture3](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/433dc2c0-8dfc-4166-b68b-d3bd7ab119c3)
![capture5](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/d8540d52-8ba9-497b-b975-115658d8d1d9)

**Comment by victokra** [comment]
@IanSB 

> The NTSC version of PETSCII robots produces 262 line video and it was added to support that, see: #266

Ah, I see. I've checked the PAL-version of that game and that still does 312 lines. The author must have chosen 262 since it's closer not NTSC.

> c0pperdragon [...] new board only works with RGBtoHDMI
> An analog RGBtoHDMI can decode that back into a C64 image. (It's a way of getting bit perfect video out of the computer on a single wire).

That's nice, though likely not for me since I got a digital RGBtoHDMI only (the one from Texelec). My C128 has a jailbar-fix and near-perfect VIC-picture already, though not HDMI...

Another issue that may fit this topic: Applying what I learned I created a 960x540-mode for the VDC (so half horizontal and vertical of FullHD) and it works nicely. However when the output resolution is set to 1920x1080 it does not scale horizontally as it should. See pictures  of source summary and geometry menu below. Do you see any reason why it would not work scale horizontally?

![IMG_6422](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/dc338d7b-cb41-41f6-88c7-575c6ab072c8)
![IMG_6423](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/e957949a-f88a-43af-b8ec-ced277ba4893)

Strange enough when I set the output resolution to 1920x1200 it scales correctly horizontally, but of course then I have a small black border and the top and bottom and this is not the native resolution of my display device.

![IMG_6421](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/7459043b-1e72-48eb-a8d0-57789095b03e)

I only get the scaling to apply when output res is set to 1920x1080 if I cange the geometry menu.

![IMG_6424](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/a87f336c-8022-4885-ab08-f2c4d61bc0b6)

Any idea what might be going on here?

**Comment by IanSB** [comment]
@victokra 
>Any idea what might be going on here?

Yes, the integer scaling was originally designed to assume 4:3 sources and scaled to fill that available area on widescreen panels.

I later introduced the H and V pixel aspect settings which allowed integer scaling to specify the aspect ratio of the LCD pixel block to be used for one source pixel which constrained the size of the resulting image and at the same time allowed a wider than 4:3 image to be displayed which was useful for displaying 768x576 cropped to 768x540 etc. 
However due to the scaling calculations that worked out at 912 pixels for a 1920x1080 and 960 pixels for 1920x1200

I have tweaked the code to use the original width of the display with integer scaling as the H and V pixel aspect should constrain the image to 4:3 for such sources so please try this:

[kernelrpi-widescreen.zip](https://github.com/hoglet67/RGBtoHDMI/files/11607451/kernelrpi-widescreen.zip)

Note that if either of the H and V aspect settings are set to 0 then it will fill the screen as much as possible. Similarly if the H and V aspect are an impossible combination like 8:1 that will also just fill the screen.

Some of the existing profiles have H and V aspect set to 0 so they should be tweaked to fix a sensible aspect ratio to avoid any issues with this change.
Note when you adjust them, a line at the top shows the current actual aspect ratio after rejecting invalid or zero settings so if that looks correct then set the values to match

Interpolated scaling has a "4:3" and "fill all" settings so you would have to select the appropriate one for those scaling modes.

**Comment by victokra** [comment]
Hooray! That works great. I checked all the profiles again. Here are the few that have been updated. I have included the one for the upcoming 960x540-mode which may prompt me to release a new version of VDC Mode Mania with the "special mode" for the RGB2HDMI. I will probably wait until the recent changes become official or at least an official beta.

Thanks for this project. I did not realize how cool the RGB2HDMI is until I was told at the Revision-demoparty to get one :)

[Commodore_128-80.zip](https://github.com/hoglet67/RGBtoHDMI/files/11611108/Commodore_128-80.zip)
![IMG_6425](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/4c3d75d2-20d8-46f2-a950-62bca0498c87)
![IMG_6426](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/5acaa955-e9d7-4ca5-8c32-7868edf8513b)

**Comment by timmcross** [comment]
> BTW I've been working with c0pperdragon to do an RGBtoHDMI adapter for the Commodore 64 video. It uses a small plug in board which sits under the VIC II to pick up the digital signals going to the VICII chip and it then sends them out of the C64 on the UHF modulator's phono in an encoded format so that an RGBtoHDMI can plug into that and decode the data to make a completely noise free bit perfect image. Do you think there would be interest in a C128 variant of that?

IMHO, I would love to give this a try for both the C64 and C128!

**Comment by xahmol** [comment]
> BTW I've been working with c0pperdragon to do an RGBtoHDMI adapter for the Commodore 64 video. > > (..)
> Do you think there would be interest in a C128 variant of that?

Would certainly also be interested, provided that it is an easy install for a hardware handicapped person like me. Does it need soldering or something like that? (Think it did not)

**Comment by IanSB** [comment]
@victokra 
@timmcross 
@xahmol 
>Does it need soldering or something like that? (Think it did not)

It can in theory be just a plug in solution. I think c0pperdragon is selling his board with crocodile clips so it can be connected to the modulator phono without soldering but that might be a problem with the breadbin C64 because the modulator lid on at least some of those models has a tab soldered to the body so getting the lid off needs a soldering iron unless the tab can be cut with some wire cutters. Also the very early breadbin model with discrete clock circuitry needs an additional socket to lift the board clear of some nearby components.

An alternative might be to use a cable with a phono dangling out the back of the C64 but that is less elegant than re-using the modulator phono. You would have to contact the sellers to see if they would offer that option.

Note you also need an analog RGBtoHDMI, either a standard one or the lower cost "lumacode" one designed for use mainly with this style of adapter 

see
https://github.com/c0pperdragon/c0pperdragon.github.io/wiki/VIC-II-dizer


also with lumacode RGBtoHDMI and case but no Pi zero:
https://www.ebay.co.uk/itm/235009314007?hash=item36b7a58cd7:g:Gi0AAOSwlIVkX0Zl

this one includes lumacode RGBtoHDMI with case, raspberry Pi & sd card)
https://www.sellmyretro.com/offer/details/63609

**Comment by victokra** [comment]
@IanSB 

I'm still experimenting with this fine device and am not sure if this is a bug or by design.

I have the 720x700 interlace mode. This is how I set it up on RGB2HDMI:

![IMG_6463](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/69cdb048-4f4c-4d80-839d-3f395f7ae887)

H Pixel Aspect = 2 and V Pixel Aspect = 3 means this is scaled to 1440 x 1050. But Min and Max V Height are both 350 as they are obviously counting the lines of each field and not both fields combined? However the real resolution in interlace is 700 lines and when this is scaled to 1050 this is factor of 1.5 and it shows in the picture I think

![IMG_6464](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/bd9e9692-2971-4167-ac97-1d6101bf26ac)

Some pixels have double height, some single height.
If I set Min V Height to 270 the picture is cut off but looks better:

![IMG_6465](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/40b7e5e2-1ca0-484e-8656-fb4c944c2c03)

All pixels have the same height again. It may be hard to see zooming in on the picture I've done with my phone.
For reference here are the capture-pictures from the device for the two settings.

![capture0](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/31010194-89cf-4b6b-819b-95d14e28c3ff)
![capture1](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/210abdcb-3827-44bf-afdf-c416e88cedb7)

Here the first picture has a height of 720 and as such does not show any distorted pixels. But when I resize this to 1080 in IrfanView for example I get a similar effect as on the 1080p-display.

So, what to make of this? Should interlaced-scaling work differently?

**Comment by IanSB** [comment]
@victokra 

>I'm still experimenting with this fine device and am not sure if this is a bug or by design.
I have the 720x700 interlace mode. This is how I set it up on RGB2HDMI:
H Pixel Aspect = 2 and V Pixel Aspect = 3 means this is scaled to 1440 x 1050. But Min and Max V Height are both 350 as they are obviously counting the lines of each field and not both fields combined? However the real resolution in interlace is 700 lines and when this is scaled to 1050 this is factor of 1.5 and it shows in the picture I think

This behaviour is by design but wasn't really intended for use with interlaced sources.

Fractional integer scaling (an integer plus a half) is allowed vertically or horizontally when double height or double width capture is enabled. (The final scaling factor is shown in the source summary page)
Double height/width means that two pixels are written to the Pi's frame buffer in the appropriate direction for each pixel captured. This has two purposes:
1. It improves scaling quality when interpolated scaling is used because for small sources like 320x200 there is less difference between the frame buffer size (e.g. 640x400) and the final display size so the image is not as soft as interpolating 320x200 directly.
2. Double height frame buffers are required for scanlines as there are two vertical pixels in the frame buffer for each source pixel.

Doubling the height and width of the frame buffer like that initially had a restricting effect on integer scaling because the frame buffer was double it's original size.

e.g. A source like a commodore 64 which is 320x200 could be integer scaled to a 1024x786 monitor with 3x scaling in both directions for 960x600 with some border.  If double height & width were both enabled, the frame buffer would be 640x400 and that doesn't integer scale at all on a 1024x768 monitor so your would be left with a 640x400 image with very large borders or have to use interpolated scaling instead.

However I realised that using 0.5 steps in integer scaling would still produce a good image because although alternate pixels would be different sizes, they would be the same colour so it wouldn't be noticeable. When the scanlines option is enabled, although the height of all the scanlines would be different to the height of all the source pixels that wouldn't be objectionable either.
e.g. in my above example of scaling 640x400 to 1024x768 with 1.5x scaling the pixels would be alternately 1 and 2 LCD pixels wide/high.

However, this doesn't work so well with interlaced sources. They make use of the double height frame buffer to store both fields as the real height of the image is twice the height in geometry but the data is different in each pair of lines so you get uneven pixels.
There is no way around this as you can't integer scale a 700x700 image to a 1920x1080 screen with the correct aspect ratio as  simply doesn't fit so you would either have to specify the H & V pixel aspect ratio of 1:2 or 2:2 to get even pixels or use interpolated scaling.
I suppose I could disable the half integer scaling when an interlaced source is detected but you can already override it as above by specifying a different aspect ratio and there are issues as the BBC micro can generate progressive images with interlaced sync and I wouldn't want to disable the half integer scaling under such conditions.

BTW 2:3 (i.e. 2 : 1.5) scaling would work on a 4K monitor

**Comment by victokra** [comment]
Thanks for the explanation. In the end it is a decision what looks best considering all constraints. I.e. do I want to display every pixel in the correct aspect ratio, but live with a small-window-boxed area? Or can I live with half-integer-scaling if the resulting shown picture is larger. Or is it possible to reduce the V height Min size just slightly, so better scaling options are availabe.

I've decided on a case-by-case basis now. The color-mode ITFLI had 1.5 scaling because it was 576 pixels high. I've reduced min-size to 540 and while loosing 36 lines it now looks much better.

On monochrome the 1.5 scaling looks ok and does not cause as much problems as with color. So I've left this for 720x700 and applied it to 800x600, so this fills a larger portion of the screen now than before.

My "approved" updated 2 profiles attached. Would be nice if you updated those 2 in the main-branch as well.

[Commodore_128-80-updated-profiles.zip](https://github.com/hoglet67/RGBtoHDMI/files/11670701/Commodore_128-80-updated-profiles.zip)

**Comment by IanSB** [comment]
@victokra 

>My "approved" updated 2 profiles attached. Would be nice if you updated those 2 in the main-branch as well.

OK I've committed that as well. Note for now this is in the dev branch on my fork where the betas are released. It will eventually find its way to the main branch here in the next stable release.
https://github.com/IanSB/RGBtoHDMI/commits/dev

**Comment by IanSB** [comment]
@victokra 
>added Cmd_128_50Hz_320.txt - there was a ROM-revision by Commodore in 1986 which adjusted PAL-settings for VDC to be more correct (15625 Hz and 312 lines). However all machines before that had slightly off PAL-timing (1016 line-length = 15748 Hz, 320 lines). I think about 50% off all PAL-C128 out there will have the older revision ROM.

I just added a Commodore 128 to my collection and this has the 320 line ROM.
However the sub-profile that often gets selected is Text-PAL_320.txt (Which has 1017 clocks per line) rather than Cmd_128_50Hz_320.txt (which has 1016 clocks per line).
The above is the only difference between them (for detection purposes) and that is too close together to be reliably differentiated so I have had to remove Text-PAL_320.txt to ensure that the standard mode works all the time.
The same is true of Cmd_128_60Hz_264.txt and Text-NTSC_264.txt so I have had to remove the latter as well.

This means your text demos using the above will not be displayed properly as they will be slightly mis-sampled.
There are a several ways around this:
I could add those as manually selectable profiles
You could modify the demos to use 1016 clocks (the standard profiles might need their max size increased)
You could change the number of lines in those demos by at least 2 (e.g. 322 and 266) which would allow updated profiles to be differentiated.

BTW my C128 had a bizarre fault:
It all worked except the 80 column output which had characters with bits 1 & 3 stuck high. I first thought this was a RAM fault so I changed VDC RAM U23 but that didn't help. I then though that the VDC might have failed but I noticed that running my fingers along the VDC pins would change the video on the screen which corresponded to touching DD1 & DD3 so they appeared to be floating and sure enough there was no continuity between those VDC lines and the DRAM.
After adding a couple of links it all started working properly. It might have been like that from new but never noticed if the 80 column mode was never used.

**Comment by victokra** [comment]
I noticed the mis-detection for the text-modes as well. These are really just add-ons to the demo, not of the utmost importance, but your post got me thinking about them again.

Effectively these modes where just designed to show how it is possible to fill our more screen area on a regular display-device.

When I originally did the demo in 2012 I wanted to keep the original Commodore-ROM-settings as much as possible, that's why they have 320 lines for PAL and 264 for NTSC. However these values were never really as close to the PAL or NTSC-standards as they could be. My intention is to make them as close to NTSC and PAL as possible, so most TVs and monitors will display them correctly.

The clock cycles need to be a multiple of 9 (coming from increasing the char-size to 9) for the wider text-screen-size. So for NTSC they can only be 1017 and for PAL they could also be 1026 to be close enough to the original horizontal frequencies for NTSC and PAL.

But then I have 262 lines on NTSC: 16.000.000 / 262 / 1017 = 60.05 Hz
and 312 lines for PAL: 16.000.000 / 312 / 1026 = 49.98 Hz
or 315 lines for PAL: 16.000.000 / 315 / 1017 = 49.94 Hz

But all these again lead to the problem of conflicts with the other (standard) sub-profiles. If I use a different number of lines this might break the modes on other display-devices which expect signals close to the PAL/NTSC-standard.

So, all things considered, option 1 " I could add those as manually selectable profiles" seems to be the most sensible solution. Is it possible to mark profiles that way, so that they can only be selected manually? I did not find such an option when fiddling around witht he modes in June.

**Comment by IanSB** [comment]
@victokra 
>Is it possible to mark profiles that way, so that they can only be selected manually?

There is an option to switch between two 'timing sets' for each profile (a timing set is all the settings in the geometry and sampling menu)
This is required for some computers that output using two different pixel clocks like the bbc micro which uses 12Mhz for teletext mode and 16Mhz for other modes. The Apple II GS also switches between 14.3Mhz and 16.3Mhz pixel clocks.
With some computers it is possible to automatically select either set 1 or 2 and those options can be chosen using the Auto Switch option. This also has a manual mode and that can be used to manually switch between the two sets. When selected you can either switch with the menu option that appears or press SW3 to toggle.

Here is a profile set with the 320 & 264 modes having set1 = normal text and set2 = demo text
[Commodore_128-80.zip](https://github.com/hoglet67/RGBtoHDMI/files/12165544/Commodore_128-80.zip)

Note that the set1/2 setting is saved so you can accidentally leave it defaulting to set2 after calibrating

I don't think this is the ideal solution as you will have to keep pressing SW3 during the demos so another option might be to have a separate VDC Demo profile with the demo versions of the 320 and 262 profiles (You could permanently use that profile set if you have the later ROMs with 312 & 262 line standard modes).

Not sure what to call such a profile, any suggestions?

Do you have a link to the latest versions of your demos as I should be able to run them now.

BTW c0pperdragon has implemented a version of his lumacode board mentioned above for the C128 which is currently being tested.

**Comment by victokra** [comment]
So, the "VDC Demo profile" would have the two sub-sets for the two text-modes? Would like to try that myself. I'd call it "VDC Mode Mania Text profile" or "VMM Text profile".

I will send you the link to the latest VDC Mode Mania as I only want to release this once at least the new beta of RGB2HDMI-firmware that supports it is out. VGA Mania you can download her:

https://www.tokra.de/c128/vdcvgamania-v101.zip

**Comment by IanSB** [comment]
@victokra 
I have released Alpha 59F here:
https://github.com/IanSB/RGBtoHDMI/issues/18#issuecomment-1650738215

Please download and try it.
There are two profiles:
Commodore_128-80
Commodore_128-80_(VDC_Text_Demos)

I have made an experimental change to the (VDC_Text_Demos) version so it should auto switch between the two 320 line and two 264 line sub-profiles. (i.e. support everything)
However this relies on the C128 16Mhz crystal being very accurate and I don't know if that is the case.

There is about 1000PPM difference between the 1016 and 1017 modes so I have reduced the Clock Tolerance setting in those 4 profiles from 5000PPM to 500PPM. As long as both the Pi zero and C128 crystals are accurate this will work. I have tried many Pi zeros and they do seem to be accurate but the C128 is more unknown.
e.g. at the moment when the standard 320 line mode is selected my PPM error is about 64PPM and switching to the 320 demo sub-profile the error changes to -908PPM which is greater than the 500 threshold so it will select the correct sub-profile.
The C128 uses an oscillator module rather than a discrete crystal circuit and they are usually trimmed for accuracy but more systems need to be tested to confirm if this is viable.
Worse case if there are inaccurate oscillators then the profile won't get selected at all and the video will flash as it can't find a match.
What sort of PPM errors do you get on your C128? (In the info menu)

**Comment by IanSB** [comment]
@victokra 
I've just been trying your demos and found some problems:
1. Spectrumania
Standard Pixel width mode works
Double pixel width mode doesn't work as there is no profile to match it (I had to add one)
That mode is outputting 320 line video with line length 1024 rather than 1016
Are you partially modifying the VDC registers as that will mean the final output timing will depend on which ROM (312 or 320) is fitted?
Would this be an issue in NTSC mode as both the 262 & 264 profiles use 1016 clock cycles?
Can't test the 4in1 as I don't have 64K (would that be any different with different ROMs?)

Here is the additional profile:

[Cmd_128_50Hz_320_1024.txt](https://github.com/hoglet67/RGBtoHDMI/files/12172582/Cmd_128_50Hz_320_1024.txt)

2. VGA mode mania
The profiles are recognised correctly but I have tweaked the "FB size" setting to Double Height + Width for both of them so the menus are more readable:

[VDC-VGA-HFLI_513.txt](https://github.com/hoglet67/RGBtoHDMI/files/12172740/VDC-VGA-HFLI_513.txt)
[VDC-VGA-Text-Mono_521.txt](https://github.com/hoglet67/RGBtoHDMI/files/12172742/VDC-VGA-Text-Mono_521.txt)

The text menu is visible but the images are garbage:
![capture12](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/5708f98c-0039-4dbe-90c4-0d32a344f023)
![capture13](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/ed70c1eb-667f-4721-a7f9-412ba3bac523)

3. VDC Mode Mania
The big text mode profiles are correctly selected with the reduced Clock tolerance setting but the images are also garbage:

![capture15](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/034e93e7-aa32-4a64-8b48-8869317618c5)
![capture5](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/fbf6f639-0304-47a4-bdae-d17cf6ffa46d)

Any idea what the problem is?
It could be a hardware issue so I will check all the address lines on the VDC DRAM but can you recommend any test programs for the C128?

**Comment by tkracke** [comment]
Will try the PPM error later tonight. @xahmol should be able to test this as well.

Regarding palettes: VDC Mode Mania and VDC VGA Mania were made for RGBI (brown) and VDC Spectrumania was made for CGA (dark yellow). I do not approve of using the "wrong" palettes for the demos ;-)

**Comment by IanSB** [comment]
@tkracke
>Regarding palettes: VDC Mode Mania and VDC VGA Mania were made for RGBI (brown) and VDC Spectrumania was made for CGA (dark yellow). I do not approve of using the "wrong" palettes for the demos ;-)

It's the other way around RGBI = dark yellow & CGA = brown.

I can leave all the VDC profiles with the correct palette and just change the palette for the standard text modes.

However there is a slight problem, these are the standard modes:
Cmd_128_50Hz_312
Cmd_128_50Hz_320
Cmd_128_60Hz_262
Cmd_128_60Hz_264

I have also added the following to cover Spectrumania:
Cmd_128_50Hz_320_1024
Cmd_128_60Hz_262_1024
Cmd_128_60Hz_264_1024
(They could be renamed to VDC-Spectrumania...)

These ones could be left permanently with the RGBI / yellow palette.

However when running Spectrumania on a PAL 312 line ROM there is no difference in the timing so Cmd_128_50Hz_312 would be selected and the palette will depend on which profile set is selected by the user.

So the two options are for the above Spectrumania profiles to always have the yellow palette with 312 line being the odd one out or for those profiles to track the user CGA/RGBI choice which would be consistent behaviour irrespective of ROM or PAL/NTSC.

Which would you prefer?

**Comment by victokra** [comment]
Ok, this is the screenshot of the source summary right after power on:

![IMG_6664](https://github.com/hoglet67/RGBtoHDMI/assets/134559359/76f0c315-2be8-4a2a-b076-097028cb8572)

I noticed in the 320 Big Text profile (choice 8 in VDC Mode Mania) the aspect ratio is 1:2 while 2:3 would look much better usually (at least at 1920x1080 which I am using).

**Comment by victokra** [comment]
For testing purposes I quickly amended VDC Mode Mania so you can choose in which ROM-setting it starts (by switching the 3 affected registers):

[vdc mode mania-t.zip](https://github.com/hoglet67/RGBtoHDMI/files/12175855/vdc.mode.mania-t.zip)

This is for the SD2IEC-version, just start this instead of the usual program, it will present with the 3 options: NTSC, PAL-320 and PAL-312.

**Comment by victokra** [comment]
>  track the user CGA/RGBI choice which would be consistent behaviour irrespective of ROM or PAL/NTSC

This option seems best. The demo itself tells you it would prefer dark yellow. So if you run it with brown you will notice and have the chance to change profiles then. Also the demo features (poor) software-emulation for dark yellow. So running this with brown active is not totally wrong.

**Comment by xahmol** [comment]
> @xahmol should be able to test this as well.

I am in holiday now and not around any Commodore for another two weeks.... so can not test on short notice.

**Comment by IanSB** [comment]
@victokra 
OK here are my proposed profile sets:
[Commodore_128_profiles.zip](https://github.com/hoglet67/RGBtoHDMI/files/12177373/Commodore_128_profiles.zip)

As discussed they are:
Commodore_128_CGA
Commodore_128_CGA_(Alt)
Commodore_128_RGBI
Commodore_128_RGBI_(Alt)

The palettes are only changed in the sub-profiles with names beginning C128_

The Clock tolerance PPM thresholds for all profiles are set to 3000 except the following in the Standard (non-Alt) profiles where they are set to 500:

C128_50Hz_320
C128_50Hz_320_1024
Text-PAL_320
C128_60Hz_264
C128_60Hz_264_1024
Text-NTSC_264

The "Text" profiles are removed in the (Alt) sets all thresholds are set to 3000 so the only thing that won't work properly with the Alt sets are those two text modes.

I also disabled the genlock in the following modes as they are below 50Hz so will never lock (They can be made to lock but then the output might not be displayed on most monitors)
VDC-HFLI_426
VDC-IM800_651i
VDC-IMONO_749i

Finally something strange happens when I select option 3 (VDC-FLI) in modemania:
All the other modes output the correct sync timing immediately they are selected during the black screen loading period but the above mode outputs some set of pulses that is being interpreted as intermittent loss of sync (the LED D2 which is normally off lights to indicate loss of sync and it is flashing) and this makes the menus unusable during that period.
Once the image is displayed the correct sync timing is output.
Is this expected behaviour?

**Comment by victokra** [comment]
Gave it a quick test, seems to work fine. Why are the (Alt)-sets needed still? Is this just for PPM-tolerance? If it is confirmed that the pixel-clock of the C128 is stable enough I would rather remove them, as this might confuse users. Also, maybe CGA and RGBI could add "brown" and "dark yellow" to the the profile-name for clarity as even I get those mixed up...

Regarding VDC-FLI: This is expected behaviour. Long answer: The 8x1 pixel mode is a hack not supported originally by the VDC. If you set the registers to 8x1-mode it will just produce a black screen. The trick is to set the mode to 8x2 in the VBLANK-area and only set it to 8x1 once the raster is outside the VBLANK. This requires an assembler loop that watches the VBLANK and I run this only once the picture has been loaded from disk fully. So, while loading the mode is not fully set up and will have more lines.

**Comment by IanSB** [comment]
@victokra 
>Why are the (Alt)-sets needed still? Is this just for PPM-tolerance? 

Yes

>If it is confirmed that the pixel-clock of the C128 is stable enough I would rather remove them, as this might confuse users

I think it is likely that the clock is stable as both our systems are less than 100PPM error and they must have been made some time apart as mine is old ROM and yours is new ROM
I will put the alt profile folders inside the other folders so they will be invisible unless manually moved up one level

>Regarding VDC-FLI: This is expected behaviour.

I just checked and the Vsync is 7.63Hz which is longer than the sync detected timeout

**Comment by IanSB** [comment]
@victokra 

I received my 4464 RAM chips today and quickly upgraded as I had already socketed the RAM during my previous fault finding so now I can run your demos.

![capture8](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/b12a06e6-f501-43ff-84fd-ce566a7ae5d9)

A couple of things:
![capture0](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/f64d17fe-943e-40d7-a94a-426b4c1747aa)

To avoid confusion, this project is "RGBtoHDMI" not "RGB2HDMI"

How long should your images take to load?
e.g. the above image for me took 3 mins 20 secs to load
I'm using a Pi1541 drive but it is in 1581 mode according to the LCD

Also is there any way of aborting a slideshow and returning to the main menu?

**Comment by victokra** [comment]
The picture is nearly full 64K, so it may take a while on the 1541. I do not know whether the Pi1541 supports burst mode or Jiffy Dos. Then it would be much faster. You can only exit by STOP/RESTORE and typing RUN.

**Comment by IanSB** [comment]
@victokra 
>I do not know whether the Pi1541 supports burst mode or Jiffy Dos. Then it would be much faster. You can only exit by STOP/RESTORE and typing RUN.

What drive emulator do you use and is it faster than my above times?

I actually developed some commercial software for the C64 back in the mid 80s but got frustrated by the slow disk speeds.
I was mainly a BBC micro programmer so I developed my software on that and designed a cartridge for the C64 that emulated a fast disk drive by linking the parallel ports on the C64 and BBC Micro so the C64 could load directly off the BBC disks.
![c64ppfs](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/74e4c8d6-7bf6-428b-876c-ba5733380506)

I've uploaded alpha59G which will be released as the next beta if no issues are found
https://github.com/IanSB/RGBtoHDMI/issues/18#issuecomment-1657223994

I did a test of all modes and found some issues:
There was no profile for VGA-TFLI so I added one:
VDC-VGA-TFLI_519
The Spectrumania 4 in 1 profile didn't work on my system as it was generating 1016 clock video rather than 1024 clocks
(Another example of the 312/320 line ROM difference.
The above release now works in all modes in all demos for me.

Can you test all modes in all three demos with your 312 line ROM and if that looks OK I will release this build

**Comment by victokra** [comment]
@IanSB 

> What drive emulator do you use and is it faster than my above times?

I ust an SD2IEC with a "hacked" firmware from 2010 or so which supports burst-mode. When doing so, each picture loads in about 5-10 seconds. Sadly most floppy emulators do not support burst mode. But you should be able to use JiffyDOS usually. The 1541 Ultimate II+ onward also support burst mode.

> (Another example of the 312/320 line ROM difference. The above release now works in all modes in all demos for me.
> Can you test all modes in all three demos with your 312 line ROM and if that looks OK I will release this build

Just tested and all modes work fine on my 312-ROM. I was wondering if I should update at least VDC Mode Mania to force the 312-line-mode for PAL? As an update to VDC Mode Mania is coming anyways, I could just force the 312-mode there. But then people with older releases might be seeing problems with the RGBtoHDMI, so maybe it is best to keep as is.

**Comment by mincebert** [comment]
I'm giving this a try on my C128DCR, using the Alpha 59 RGBtoHDMI software — all three demos seem to work OK (VDC Mode Mania, VDC Spectrumania and VDC VGA Mania), albeit with some caveats...

The Clock Error in all modes is -377 PPM, with the exception of some of the Mode Mania demos, where it is slightly different (e.g. -370, -376 and I think I saw a -380).

When running the VGA Mania demo, none of the modes seem to be selected automatically: it starts up with me staying in the original C128 50Hz 312 mode and the screen flashing.  If I manually change to VDC-VGA-Text-Mono 521, I get a stable picture but off to the right a few characters and have to bring it back left with the cursor keys.  I've tried moving it left without first changing the subprofile and it still doesn't lock on.  When running the HFLI and TFLI demos in that, I again need to manually select the subprofile.  Once selected, though, it's perfectly stable.

 I have a 312 line ROM but my machine is of an odd providence: it was originally a French machine but I replaced the Character ROM with a US one and did a modification of the jumpers controlling the font selection, ROM and 40/80 key.  It's possible I've messed something up somewhere!

A couple of screen captures of the Summary screen.  The 50Hz 312 line mode on startup:

![50Hz 312](https://github.com/hoglet67/RGBtoHDMI/assets/21294069/2d6d05a9-fea1-4a4f-bc3c-479e1bcb51bb)

And the VGA Text Mono mode; the screen has been shifted to the left to get all the characters on:

![VGA Text Mono](https://github.com/hoglet67/RGBtoHDMI/assets/21294069/b68abf13-ebb4-4784-8901-9eb7bef05689)

**Comment by IanSB** [comment]
@mincebert 

The H and V timing on your C128DCR is correct as both the clock error and lines per frame are OK but the syncs are inverted compared to mine (-H -V instead of +H +V) so the profile match fails.
Here is my source summary for comparison:
![capture2](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/739cecf9-35e0-4614-b24b-69627d04893f)

I think -H-V syncs are the correct polarity for VGA so I guess my earlier machine can't support sync inversion 

Here are three additional VGA profiles with the sync inverted:
[VDC-VGA-H-V.zip](https://github.com/hoglet67/RGBtoHDMI/files/12227472/VDC-VGA-H-V.zip)

Unzip those into:
\Profiles\6-12_BIT_RGB\Commodore\Commodore_128_CGA_(Brown)
and 
\Profiles\6-12_BIT_RGB\Commodore\Commodore_128_RGBI_(Yellow)

Then reboot and see if that makes any difference

@victokra 

Do you know anything about the sync inversion capability of the C128 or C128DCR?
I looked at the DCR schematic and there is no external circuitry in the sync path so is this a difference between revisions of the VDC chip?

Also mincebert's PPM error is quite large so there is a possibility that some systems will exceed the 500PPM threshold
This may only be applicable to the C128DCR as that uses a 16Mhz crystal rather than a 16Mhz oscillator module and that would not be as accurate as an oscillator module which might have been trimmed for accuracy during manufacture.
See pcb photo here:
https://myoldcomputer.nl/wp-content/uploads/2016/01/mainpcb-250477.jpg

If it's only going to affect the C128DCR and being later, those systems are always 312 line ROM then it is not so much of an issue as the 312 line profiles have a 3000PPM threshold so only the big text modes would fail. It is only the 320 & 264 line profiles that have a 500PPM threshold and that would cause the main text mode as well as the big text modes to fail.

>I was wondering if I should update at least VDC Mode Mania to force the 312-line-mode for PAL? As an update to VDC Mode Mania is coming anyways, I could just force the 312-mode there. But then people with older releases might be seeing problems with the RGBtoHDMI, so maybe it is best to keep as is.

Forcing 312 line for the big text modes would be a good idea as that would allow for much wider thresholds and allow for the option to revise the profiles if the 500PPM threshold becomes an issue. (The profiles could support both for now)

What happens to these big text modes on an NTSC system?
Is there a difference between old and new ROMs in NTSC (I already had 264 and 262 line profiles based on user feedback)

**Comment by victokra** [comment]
@IanSB 

> the syncs are inverted compared to mine (-H -V instead of +H +V) so the profile match fails. 
> think -H-V syncs are the correct polarity for VGA so I guess my earlier machine can't support sync inversion
> Do you know anything about the sync inversion capability of the C128 or C128DCR? I looked at the DCR schematic and
> there is no external circuitry in the sync path so is this a difference between revisions of the VDC chip?

Oh yeah, the newer VDC (8568) in the C128DCR has a 37th register that can set sync-polarity that the older VDC (8563) does not have. For compatibility reasons I set the 37th register to VGA. As the old VDC does not support this register the write just goes nowhere. So I guess you should add profiles with correct sync-polarity for this as well.

> I was wondering if I should update at least VDC Mode Mania to force the 312-line-mode for PAL?

> Forcing 312 line for the big text modes would be a good idea [...]

Ok, will create an updated version later today and post it here.

> What happens to these big text modes on an NTSC system? Is there a difference between old and new ROMs in NTSC
> (I already had 264 and 262 line profiles based on user feedback)

The NTSC-values never changed in the ROM-revisions. They are always 264 lines. The 262-line setting is only for Attack of the PETSCII-robots which changed this setting since it is actually closer to NTSC. Maybe this could be reflected in the profile name?

**Comment by IanSB** [comment]
@victokra 

>Ok, will create an updated version later today and post it here.
I would do the updates as laid out in earlier post:
262 lines on NTSC: 16.000.000 / 262 / 1017 = 60.05 Hz
312 lines on PAL: 16.000.000 / 312 / 1026 = 49.98 Hz

Actually that would create the same sort of conflict with the standard modes:
262 / 1016 (500PPM)
312 / 1024 (1000PPM)

Could you change to some other lines per frame?
This has to be + or - 2 lines from either 262/264 ( so 260 or 266) or 320/312 (so 314 to 318).
Note +-1 is not possible as that counts as a match. 
The existing demos have ranged from 312 to 320 so a minor change from those is not going to matter.

**Comment by victokra** [comment]
Ok, the changes I initially had in mind would be the nearest to NTSC/PAL-standard but as these might conflict with existing modes this seems not to be the best option. I am now just going to make it so that the 320-machines force the 312-settings.

**Comment by IanSB** [comment]
@victokra 
 >I am now just going to make it so that the 320-machines force the 312-settings.

Ok that should fix the conflict on 312 line machines as the profiles will be:
312 / 1017 vs 312 / 1024 which is ~3500PPM

There is still a 500PPM conflict between standard NTSC text mode and Big Text NTSC:
264 / 1016 vs 264 / 1017

The only way around that is to change the line count to something other than 262 or 264 or change the clocks from 1017 to something else like 1026

**Comment by victokra** [comment]
> There is still a 500PPM conflict between standard NTSC text mode and Big Text NTSC: 264 / 1016 vs 264 / 1017
> 
> The only way around that is to change the line count to something other than 262 or 264 or change the clocks from
> 1017 to something else like 1026

That would mean 260 or 266 since +-1 ist not possible and that would be too far off NTSC-standard and could lead to problems with other display devices. Same for changing the horizontal frequency. Older display devices are even more picky about that. So I'd rather leave NTSC as is.

**Comment by victokra** [comment]
Forget what I said earlier. I checked my source again and actually I always set the three registers 0,4 and 7 that differ between the ROM-revisions and PAL/NTSC explicitly for each mode in VDC Mode Mania. So for that demo at least it should not matter which ROM or TV-standard your machine has.

Should I revise VGA Mania or Spectrumania in the future I will amend this issue, but for now there is really nothing else that would warrant a new release.

I revised the texts to "RGBtoHDMI" and while at it implemented an (undocumented) option to exit the viewer with ESC so you do not have to go through all the pictures. Grab the most updated version here:

https://www.tokra.de/mike/vdcmodemania-v12.zip

**Comment by IanSB** [comment]
@victokra 
OK they all seem to work for me.

Due to the very likely possibility of some C128DCRs exceeding 500PPM I have decided to include the alternate profile sets but I have named them to provide the hint of when to use them and they are also located at the bottom of the list of profiles:
![capture0](https://github.com/hoglet67/RGBtoHDMI/assets/45300594/df04c6a6-e627-494e-9d6c-046c43bad72e)

As previously mentioned, the Alt profiles have the Big text Pal and Ntsc profiles removed and all the remaining profiles increased to 3000PPM. 
All images should work without flashing screens but the big text images will have a column of noise as they are being mis-sampled by one clock.

The only remaining thing that would be nice is to get someone with an NTSC machine to test it all.

**Comment by victokra** [comment]
There are a few other VDC Demos that use different resolutions, like "Graphic Booster" or this one from 1995:

https://csdb.dk/release/?id=44110

Or this little thing from me which again uses VGA-frequencies and produces a VIC-20 like screen:

http://www.tokra.de/c128/vgavic20.zip

Should be easy enough to create profiles for this as well. Not sure how much profiles you want to include. This could become a lot as there is more software out there which messes with the VDC.

**Comment by mincebert** [comment]
> @mincebert
> 
> The H and V timing on your C128DCR is correct as both the clock error and lines per frame are OK but the syncs are inverted compared to mine (-H -V instead of +H +V) so the profile match fails. ...
> 
> Here are three additional VGA profiles with the sync inverted: [VDC-VGA-H-V.zip](https://github.com/hoglet67/RGBtoHDMI/files/12227472/VDC-VGA-H-V.zip)

Thanks — I've tried these additional profiles and the three VGA modes are now detected automatically.  They also appear centred, without me having to shift things into the middle.

I've tried all the other modes and everything all seems to be selected and display OK, without fiddling about (aside from some auto calibration).

**Comment by IanSB** [comment]
@victokra 
@mincebert 
Beta 59 release here:
https://github.com/IanSB/RGBtoHDMI/releases

I will look at adding further profiles to the Commodore 128 later but at the moment I'm working on an option to automatically create profiles using some basic information like the resolution and clock frequency plus the detected sync timing.

**Comment by IanSB** [comment]
@victokra 

I've released a new Alpha61D here:
https://github.com/IanSB/RGBtoHDMI/issues/21#issuecomment-1774182869

This has a new profile creation menu which should make profile creation easier so perhaps you could give that a try to see if you can create profiles for some of the other demo modes not yet supported.
There is some help in the info menu and in the create profile menu itself.

**Comment by victokra** [comment]
Gave this a quick test-run with an obscure VDC demo and it worked fine. This should make it a LOT easier to create new profiles in the future, especially just the ones needed to run certain software, which need not be 100% correct all the time. Maybe this process could even be automatic by the RGBtoHDMI, i.e. if no fitting profile is found the device automatically tries to create one on the fly and just pops up a quick-info box that it has just done so and if you want to save it.

**Comment by IanSB** [comment]
@victokra 
>if no fitting profile is found the device automatically tries to create one on the fly

I think this might work for the C128 but wouldn't be as useful for other random sources as they would almost certainly be mis-sampled and noisy. Perhaps just prompting to enter the create profile menu after a few seconds of no match would be a more general solution.

As long as you have a valid sync source connected the analyse timing option should produce a locked image but the hardest part of setting up a profile is getting the clock / line length combination right. There are many combinations of clock and line length that will fit a given horizontal line time but only one combination is correct. If you get that combination wrong you will get a mis-sampled image so determining either the clock or line length initially is important (the other one is calculated from the timing using clock period x line length = measured line time). 

The C128 has a fixed 16Mhz clock so analyse timing should in most cases produce a usable solution if your base profile is already a C128 one with the clock preset to 16Mhz although the pixel dimensions really need to be determined as well. 

For an unknown system with undetermined clock and line length there isn't really an easy way to automate that with the current hardware although maybe I could try to auto detect the columns of noise that are used as visual cues when manually adjusting the line length.

BTW did you find the help pages in the create profile menu sufficient or do you think anything important is missing?
I will expand on that help in the wiki eventually but I want the basic help in the menu to be sufficient for someone to have a go at creating a profile.

**Comment by victokra** [comment]
I think the help pages are sufficent and give a good first overview of what is needed. Maybe there could be some paragraphs in there or a numbered list to make the steps needed more clear and distinguishable.

**Comment by IanSB** [comment]
@victokra 
beta 61 now available:
https://github.com/IanSB/RGBtoHDMI/releases
